dbms/src/Flash/Mpp/FineGrainedShuffleWriter.cpp

```diff
@@ -0,0 +1,162 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <Common/TiFlashException.h>
+#include <Flash/Coprocessor/CHBlockChunkCodec.h>
+#include <Flash/Mpp/FineGrainedShuffleWriter.h>
+#include <Flash/Mpp/HashBaseWriterHelper.h>
+#include <Flash/Mpp/MPPTunnelSet.h>
+
+namespace DB
+{
+template <class StreamWriterPtr>
+FineGrainedShuffleWriter<StreamWriterPtr>::FineGrainedShuffleWriter(
+    StreamWriterPtr writer_,
+    std::vector<Int64> partition_col_ids_,
+    TiDB::TiDBCollators collators_,
+    bool should_send_exec_summary_at_last_,
+    DAGContext & dag_context_,
+    uint64_t fine_grained_shuffle_stream_count_,
+    UInt64 fine_grained_shuffle_batch_size_)
+    : DAGResponseWriter(/*records_per_chunk=*/-1, dag_context_)
+    , should_send_exec_summary_at_last(should_send_exec_summary_at_last_)
+    , writer(writer_)
+    , partition_col_ids(std::move(partition_col_ids_))
+    , collators(std::move(collators_))
+    , fine_grained_shuffle_stream_count(fine_grained_shuffle_stream_count_)
+    , fine_grained_shuffle_batch_size(fine_grained_shuffle_batch_size_)
+{
+    rows_in_blocks = 0;
+    partition_num = writer_->getPartitionNum();
+    RUNTIME_CHECK(dag_context.encode_type == tipb::EncodeType::TypeCHBlock);
+    chunk_codec_stream = std::make_unique<CHBlockChunkCodec>()->newCodecStream(dag_context.result_field_types);
+}
+
+template <class StreamWriterPtr>
+void FineGrainedShuffleWriter<StreamWriterPtr>::finishWrite()
+{
+    if (should_send_exec_summary_at_last)
+    {
+        batchWriteFineGrainedShuffle<true>();
+    }
+    else
+    {
+        batchWriteFineGrainedShuffle<false>();
+    }
+}
+
+template <class StreamWriterPtr>
+void FineGrainedShuffleWriter<StreamWriterPtr>::write(const Block & block)
+{
+    if (block.columns() != dag_context.result_field_types.size())
+        throw TiFlashException("Output column size mismatch with field type size", Errors::Coprocessor::Internal);
+    size_t rows = block.rows();
+    rows_in_blocks += rows;
+    if (rows > 0)
+    {
+        blocks.push_back(block);
+    }
+
+    if (static_cast<UInt64>(rows_in_blocks) >= fine_grained_shuffle_batch_size)
+        batchWriteFineGrainedShuffle<false>();
+}
+
+template <class StreamWriterPtr>
+template <bool send_exec_summary_at_last>
+void FineGrainedShuffleWriter<StreamWriterPtr>::writePackets(std::vector<TrackedMppDataPacket> & packets)
+{
+    size_t part_id = 0;
+
+    if constexpr (send_exec_summary_at_last)
+    {
+        tipb::SelectResponse response;
+        addExecuteSummaries(response, /*delta_mode=*/false);
+        /// Sending the response to only one node, default the first one.
+        assert(!packets.empty());
+        packets[0].serializeByResponse(response);
+        writer->write(packets[0].getPacket(), 0);
+        part_id = 1;
+    }
```

Comment:
There is only one packet that needs to hold the execution summary.
`HashPartitionWriter::writePackets` the same.